core/src/banking_simulation.rs

```diff
@@ -0,0 +1,616 @@
+                        let non_vote_tps = (non_vote_tx_count - last_non_vote_count) as f64 / duration;
+                        let tpu_vote_tps = (tpu_vote_tx_count - last_tpu_vote_tx_count) as f64 / duration;
+                        let gossip_vote_tps = (gossip_vote_tx_count - last_gossip_vote_tx_count) as f64 / duration;
+                        info!(
+                            "senders(non-,tpu-,gossip-vote): tps: {} (={}+{}+{}) over {:?} not-recved: ({}+{}+{})",
+                            tps as u64,
+                            non_vote_tps as u64,
+                            tpu_vote_tps as u64,
+                            gossip_vote_tps as u64,
+                            log_interval,
+                            non_vote_sender.len(),
+                            tpu_vote_sender.len(),
+                            gossip_vote_sender.len(),
+                        );
+                        last_log_duration = simulation_duration_since_base;
+                        last_tx_count = current_tx_count;
+                        (last_non_vote_count, last_tpu_vote_tx_count, last_gossip_vote_tx_count) = (
+                            non_vote_tx_count, tpu_vote_tx_count, gossip_vote_count
+                        );
+                    }
+
+                    if exit.load(Ordering::Relaxed) {
+                        break;
+                    }
+                }
+                info!(
+                    "terminating to send...: non_vote: {} ({}), tpu_vote: {} ({}), gossip_vote: {} ({})",
+                    non_vote_count,
+                    non_vote_tx_count,
+                    tpu_vote_count,
+                    tpu_vote_tx_count,
+                    gossip_vote_count,
+                    gossip_vote_tx_count
+                );
+                // hold these senders in join_handle to control banking stage termination!
+                (non_vote_sender, tpu_vote_sender, gossip_vote_sender)
+            }
+        })?;
+
+        sleep(warmup_duration);
+        info!("warmup done!");
+
+        loop {
+            if poh_recorder.read().unwrap().bank().is_none() {
+                let next_leader_slot = leader_schedule_cache.next_leader_slot(
+                    &simulated_leader,
+                    bank.slot(),
+                    &bank,
+                    Some(&blockstore),
+                    GRACE_TICKS_FACTOR * MAX_GRACE_SLOTS,
+                );
+                debug!("{next_leader_slot:?}");
+                poh_recorder
+                    .write()
+                    .unwrap()
+                    .reset(bank.clone_without_scheduler(), next_leader_slot);
+                info!("Bank::new_from_parent()!");
+
+                let old_slot = bank.slot();
+                if let Some(event_time) =
+                    self.banking_trace_events.freeze_time_by_slot.get(&old_slot)
+                {
+                    if log_enabled!(log::Level::Info) {
```

Comment:
logging functionality here should really get separated, it is quite long and distracting from the behavior of the loop.